flink-runtime/src/test/java/org/apache/flink/runtime/deployment/ShuffleDescriptorTest.java

```diff
@@ -0,0 +1,187 @@
+import org.apache.flink.runtime.io.network.partition.ResultPartitionType;
+import org.apache.flink.runtime.jobgraph.IntermediateDataSetID;
+import org.apache.flink.runtime.shuffle.NettyShuffleDescriptor;
+import org.apache.flink.runtime.shuffle.NettyShuffleMaster;
+import org.apache.flink.runtime.shuffle.PartitionDescriptor;
+import org.apache.flink.runtime.shuffle.ProducerDescriptor;
+import org.apache.flink.runtime.shuffle.ShuffleDescriptor;
+import org.apache.flink.runtime.shuffle.UnknownShuffleDescriptor;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import javax.annotation.Nullable;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.concurrent.ExecutionException;
+
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for the {@link ShuffleDescriptor}.
+ */
+public class ShuffleDescriptorTest extends TestLogger {
+	private static final ConnectionID CONNECTION =
+		new ConnectionID(new InetSocketAddress(InetAddress.getLoopbackAddress(), 5000), 0);
+
+	/**
+	 * Tests the deployment descriptors for local, remote, and unknown partition
+	 * locations (with lazy deployment allowed and all execution states for the
+	 * producers).
+	 */
+	@Test
+	public void testMixedLocalRemoteUnknownDeployment() throws Exception {
+		ResourceID consumerResourceID = ResourceID.generate();
+
+		// Local and remote channel are only allowed for certain execution
+		// states.
+		for (ExecutionState state : ExecutionState.values()) {
+			ResultPartitionID localPartitionId = new ResultPartitionID();
+			ResultPartitionDeploymentDescriptor localPartition =
+				createResultPartitionDeploymentDescriptor(localPartitionId, consumerResourceID);
+
+			ResultPartitionID remotePartitionId = new ResultPartitionID();
+			ResultPartitionDeploymentDescriptor remotePartition =
+				createResultPartitionDeploymentDescriptor(remotePartitionId, ResourceID.generate());
+
+			ResultPartitionID unknownPartitionId = new ResultPartitionID();
+
+			ShuffleDescriptor localShuffleDescriptor =
+				getConsumedPartitionShuffleDescriptor(localPartitionId, state, localPartition, true);
+			ShuffleDescriptor remoteShuffleDescriptor =
+				getConsumedPartitionShuffleDescriptor(remotePartitionId, state, remotePartition, true);
+			ShuffleDescriptor unknownShuffleDescriptor =
+				getConsumedPartitionShuffleDescriptor(unknownPartitionId, state, null, true);
+
+			// These states are allowed
+			if (state == ExecutionState.RUNNING ||
+				state == ExecutionState.FINISHED ||
+				state == ExecutionState.SCHEDULED ||
+				state == ExecutionState.DEPLOYING) {
+				NettyShuffleDescriptor nettyShuffleDescriptor;
+
+				// Create local or remote channels
+				assertThat(localShuffleDescriptor, instanceOf(NettyShuffleDescriptor.class));
```

Comment:
We could introduce two helper methods to verify the `NettyShuffleDescriptor` and `UnknownShuffleDescriptor` if you like:

```
private void verifyNettyShuffleDescriptor(ShuffleDescriptor descriptor) {
     assertThat(descriptor, instanceOf(NettyShuffleDescriptor.class));
     assertThat(descriptor.isUnknown(), is(false));
}
```

Then we could reduce many common lines in these two tests.